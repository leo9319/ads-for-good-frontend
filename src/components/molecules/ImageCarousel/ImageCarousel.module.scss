@import '../../../styles/breakpoints.scss';

.container {
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;

  &:has(> .thumbnailMode) {
    margin-block: 0px;
  }

  > div {
    display: flex;
  }

  .thumbnailMode {
    &.next,
    &.previous {
      width: 20px;
      height: 100%;

      button {
        visibility: hidden;
        padding: 20px !important;
      }

      &:hover {
        button {
          visibility: visible !important;
        }
      }
    }

    &.previous button {
      margin-left: -35px;
    }
  }

  .next,
  .previous {
    position: absolute;
    z-index: 10;
  }

  .next {
    right: 10px;
  }

  .previous {
    left: 10px;
  }

  @media (max-width: $xs-max) {
    .next {
      right: 0px;
    }

    .previous {
      left: 0px;
    }

    .next,
    .previous {
      height: 80vh;

      button {
        padding: 15vh !important;
        background-color: transparent !important;
        color: transparent !important;
      }

      button:active,
      button:focus {
        background-color: transparent !important;
        box-shadow: none !important;
      }
    }

    .page {
      bottom: 5%;
    }
  }

  .slider {
    width: 100%;
    height: 100%;
    text-align: center;
    overflow: hidden;
  }

  .slides {
    display: flex;

    > div.slide {
      margin-right: 15px;
      background-image: var(--slide-image);
      background-size: cover;
      background-position: center;
      background-repeat: no-repeat;
      width: 120.5vh;
      height: 80.2vh;
      border-radius: 24px;

      &.thumbnailMode {
        width: 80px;
        height: 80px;
        border-radius: 40px;
      }
    }

    .tooltip {
      position: relative;
      z-index: 15;

      .tooltipText {
        visibility: hidden;
        align-self: stretch;
        background-color: #ffffff;
        color: #171923;
        font-family: var(--typography-default-font-family);
        font-size: var(--typography-font-size-2);
        font-style: normal;
        font-weight: var(--typography-font-weight-regular);
        line-height: var(--typography-line-height-2);
        padding: 0;
        margin: 0;
        border-radius: 6px;
        position: absolute;
        width: 120px;
        bottom: 55px;
        right: 20px;
      }

      button {
        position: absolute;
        right: 25px;
        bottom: 15px;

        @media (max-width: $xs-max) {
          padding: var(--spacing-2xs) !important;
        }
      }

      button:focus ~ .tooltipText {
        visibility: visible;
      }
    }
  }

  .page {
    cursor: pointer;
    background: #17191f;
    text-decoration: none;
    border-radius: 15px;
    padding-block: 1vh;
    padding-inline: 2vh;
    display: inline-flex;
    position: absolute;
    bottom: 9%;
    font-family: var(--typography-default-font-family);
    font-style: normal;
    font-weight: var(--typography-font-weight-regular);
    font-size: var(--typography-font-size-2);
    line-height: var(--typography-line-height-1);
    text-align: center;
    color: #ffffff;

    @media (max-width: $xs-max) {
      bottom: 4%;
    }
  }
}
