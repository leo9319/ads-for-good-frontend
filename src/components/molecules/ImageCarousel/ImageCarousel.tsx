import React, { ReactElement, useEffect, useRef, useState } from 'react';

import Button from '@radix-styles/atoms/Button';
import { classNames } from '@utils/common/classNames';

import { ImageSlide } from './components/ImageSlide';

import styles from './ImageCarousel.module.scss';
import { AnalyticsHandler } from '@utils/Analytics';

export interface SlideProp {
  /**
   * Image source for the slide
   * @property
   */
  src: string;
  /**
   * Tooltip for the slide, if exists
   * @property
   */
  tooltip?: string;
  /**
   * Tooltip title for the slide, if exists
   * @property
   */
  tooltipTitle?: string;
}

export interface ImageCarouselProps {
  /**
   * List of slides to display
   * @property
   */
  slides: SlideProp[];
  /**
   * Carousel mode normal or thumbnail view
   * @property
   */
  mode?: 'default' | 'thumbnail';
  /**
   * Maximum width of the carousel
   * @property
   * @default '100%'
   */
  maxWidth?: string;
}

export const ImageCarousel = ({
  slides = [],
  mode = 'default',
  maxWidth,
}: ImageCarouselProps): ReactElement => {
  const ref = useRef<HTMLDivElement>(null);
  const [active, setActive] = useState(0);
  const { trackClick } = AnalyticsHandler();

  const slidesLength = slides.length;
  const next = active === slidesLength ? 1 : active + 1;
  const previous = active === 1 ? slidesLength : active - 1;

  const goToActive = (index: number) => {
    const element = document.scrollingElement ?? document.body;
    const y = element.scrollTop;
    const x = element.scrollLeft;
    document.getElementById('slide-' + index)?.scrollIntoView({
      behavior: 'auto',
      block: 'nearest',
      inline: 'center',
    });
    element.scrollTo(x, y);
  };

  useEffect(() => {
    goToActive(active);
  }, [active]);

  useEffect(() => {
    setActive(1);
    const resizeObserver = new ResizeObserver(() => {
      goToActive(1);
    });

    if (ref.current) {
      resizeObserver.observe(ref.current);
    }

    return () => {
      if (ref.current) {
        resizeObserver.unobserve(ref.current);
      }
    };
  }, []);

  useEffect(() => {
    // handle keyboard events. TODO - Need to support the enter & space for respective buttons
    const handleKeyDown = (event: { key: string }) => {
      if (event.key === 'ArrowRight') {
        setActive(next);
      } else if (event.key === 'ArrowLeft') {
        setActive(previous);
      }
    };

    window.addEventListener('keydown', handleKeyDown);
    return () => {
      window.removeEventListener('keydown', handleKeyDown);
    };
  }, [active]);

  // Extra slide for Edge Preview of the first and last slide
  const firstSlide = slides.slice(slides.length - 1)[0];
  const lastSlide = slides.slice(0, 1)[0];

  // mini thumbnail slider
  const thumbnailMode = mode === 'thumbnail';
  const defaultMode = !thumbnailMode;
  const thumbnail = thumbnailMode ? styles.thumbnailMode : null;
  const previousButtonName = 'Carousel Previous Button';
  const nextButtonName = 'Carousel Next Button';

  const track = (name: string, index: number) => {
    trackClick({
      name: name,
      contentTypeOrPosition: index.toString(),
      type: 'carousel',
      eventDetails: 'carousel click',
    });
  };
  const previousButtonClick = () => {
    setActive(previous);
    track(previousButtonName, previous);
  };
  const nextButtonClick = () => {
    setActive(next);
    track(nextButtonName, next);
  };

  return (
    <div ref={ref} className={styles.container} style={{ maxWidth }}>
      <div className={classNames(styles.previous, thumbnail)}>
        {defaultMode ? (
          <Button
            ariaLabelName={previousButtonName}
            mode="tertiary"
            size="xl"
            icon="icon-chevron-left"
            onPointerDown={previousButtonClick}
          />
        ) : (
          <Button
            ariaLabelName={previousButtonName}
            mode="tertiary"
            size="sm"
            icon="icon-chevron-left"
            rounded
            onPointerDown={previousButtonClick}
          />
        )}
      </div>
      <div className={styles.slider}>
        <div className={styles.slides}>
          <ImageSlide
            index={0}
            {...firstSlide}
            styles={styles}
            thumbnailMode={thumbnailMode}
          />
          {slides.map((slide, index) => (
            <ImageSlide
              key={slide.src + '-' + index}
              index={index + 1}
              {...slide}
              styles={styles}
              thumbnailMode={thumbnailMode}
            />
          ))}
          <ImageSlide
            index={slidesLength + 1}
            {...lastSlide}
            styles={styles}
            thumbnailMode={thumbnailMode}
          />
        </div>
      </div>
      <div className={classNames(styles.next, thumbnail)}>
        {defaultMode ? (
          <Button
            ariaLabelName={nextButtonName}
            mode="tertiary"
            size="xl"
            icon="icon-chevron-right"
            onPointerDown={nextButtonClick}
          />
        ) : (
          <Button
            ariaLabelName={nextButtonName}
            mode="tertiary"
            size="sm"
            icon="icon-chevron-right"
            rounded
            onPointerDown={nextButtonClick}
          />
        )}
      </div>
      {defaultMode && (
        <div className={styles.page}>
          {active}/{slidesLength}
        </div>
      )}
    </div>
  );
};

export default ImageCarousel;
