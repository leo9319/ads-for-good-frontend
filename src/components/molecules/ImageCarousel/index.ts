export { default } from './ImageCarousel';
export { ImageCarousel } from './ImageCarousel';
export type { ImageCarouselProps, SlideProp } from './ImageCarousel';
