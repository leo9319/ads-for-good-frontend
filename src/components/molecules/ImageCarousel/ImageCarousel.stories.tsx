import type { Meta, StoryObj } from '@storybook/react';
import ImageCarousel from '@components/molecules/ImageCarousel';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Library/Molecules/ImageCarousel',
  component: ImageCarousel,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'fullscreen',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  argTypes: {
    slides: {
      description: 'Array of slides to display',
      control: {
        type: 'object',
      },
    },
    mode: {
      description: 'Carousel Mode',
      options: ['default', 'thumbnail'],
      control: { type: 'select' },
    },
  },
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
} satisfies Meta<typeof ImageCarousel>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args

export const defaultMode: Story = {
  args: {
    slides: [
      {
        src: 'https://dummyimage.com/872x470/000/fff&text=1+Image',
        tooltipTitle: '1',
        tooltip: 'Tooltip 1',
      },
      {
        src: 'https://dummyimage.com/872x470/122/fff&text=2+Image',
        tooltipTitle: '2',
        tooltip: 'Tooltip 2 Tooltip 2',
      },
      {
        src: 'https://dummyimage.com/872x470/222/fff&text=3+Image',
        tooltipTitle: '3',
        tooltip: 'Tooltip 3 Tooltip 3 Tooltip 3',
      },
      {
        src: 'https://dummyimage.com/872x470/999/fff&text=4+Image',
        tooltipTitle: '4',
        tooltip:
          'Tooltip 4 Tooltip 4 Tooltip 4 Tooltip 4 Tooltip 4 Tooltip 4 Tooltip 4 Tooltip 4',
      },
      {
        src: 'https://dummyimage.com/872x470/300/fff&text=5+Image',
        tooltipTitle: '5',
        tooltip: 'Tooltip 5',
      },
    ],
    mode: 'default',
  },
};

export const thumbnailMode: Story = {
  args: {
    slides: [
      {
        src: 'https://dummyimage.com/872x470/000/fff&text=1+Image',
      },
      {
        src: 'https://dummyimage.com/872x470/122/fff&text=2+Image',
      },
      {
        src: 'https://dummyimage.com/872x470/222/fff&text=3+Image',
      },
      {
        src: 'https://dummyimage.com/872x470/999/fff&text=4+Image',
      },
      {
        src: 'https://dummyimage.com/872x470/300/fff&text=5+Image',
      },
    ],
    mode: 'thumbnail',
    maxWidth: '400px',
  },
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'lightgrey',
      values: [{ name: 'lightgrey', value: '#d3d3d3' }],
    },
  },
};
