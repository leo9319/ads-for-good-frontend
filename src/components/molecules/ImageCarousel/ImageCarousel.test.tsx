import React from 'react';
import { render } from '@testing-library/react';
import ImageCarousel from './ImageCarousel';

// Mock
window.HTMLElement.prototype.scrollIntoView = jest.fn();
window.HTMLElement.prototype.scrollTo = jest.fn();
globalThis.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

describe('ImageSlide', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <ImageCarousel
        slides={[
          {
            src: 'https://dummyimage.com/872x470/000/fff&text=1+Image',
            tooltipTitle: '1',
            tooltip: 'Tooltip 1',
          },
          {
            src: 'https://dummyimage.com/872x470/122/fff&text=2+Image',
            tooltipTitle: '2',
            tooltip: 'Tooltip 2',
          },
          {
            src: 'https://dummyimage.com/872x470/222/fff&text=3+Image',
            tooltipTitle: '3',
            tooltip: 'Tooltip 3',
          },
          {
            src: 'https://dummyimage.com/872x470/999/fff&text=4+Image',
            tooltipTitle: '4',
            tooltip: 'Tooltip 4',
          },
          {
            src: 'https://dummyimage.com/872x470/300/fff&text=5+Image',
            tooltipTitle: '5',
            tooltip: 'Tooltip 5',
          },
        ]}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
