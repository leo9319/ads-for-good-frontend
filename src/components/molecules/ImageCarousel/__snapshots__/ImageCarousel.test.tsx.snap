// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSlide should render successfully 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <div
        class="previous"
      >
        <button
          aria-label="Carousel Previous Button"
          class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton xl tertiary none"
          data-accent-color=""
        >
          <div
            class="buttonContent iconLeft iconOnly"
          >
            <span
              class="icon"
              style="border-radius: 0;"
            >
              <span
                class="icon icon-previous"
              />
            </span>
          </div>
        </button>
      </div>
      <div
        class="slider"
      >
        <div
          class="slides"
        >
          <div
            class="slide"
            id="slide-0"
            style="--slide-image: url(https://dummyimage.com/872x470/300/fff&text=5+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                5
              </h5>
              <p>
                Tooltip 5
              </p>
            </div>
          </div>
          <div
            class="slide"
            id="slide-1"
            style="--slide-image: url(https://dummyimage.com/872x470/000/fff&text=1+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                1
              </h5>
              <p>
                Tooltip 1
              </p>
            </div>
          </div>
          <div
            class="slide"
            id="slide-2"
            style="--slide-image: url(https://dummyimage.com/872x470/122/fff&text=2+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                2
              </h5>
              <p>
                Tooltip 2
              </p>
            </div>
          </div>
          <div
            class="slide"
            id="slide-3"
            style="--slide-image: url(https://dummyimage.com/872x470/222/fff&text=3+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                3
              </h5>
              <p>
                Tooltip 3
              </p>
            </div>
          </div>
          <div
            class="slide"
            id="slide-4"
            style="--slide-image: url(https://dummyimage.com/872x470/999/fff&text=4+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                4
              </h5>
              <p>
                Tooltip 4
              </p>
            </div>
          </div>
          <div
            class="slide"
            id="slide-5"
            style="--slide-image: url(https://dummyimage.com/872x470/300/fff&text=5+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                5
              </h5>
              <p>
                Tooltip 5
              </p>
            </div>
          </div>
          <div
            class="slide"
            id="slide-6"
            style="--slide-image: url(https://dummyimage.com/872x470/000/fff&text=1+Image);"
          />
          <div
            class="tooltip"
          >
            <button
              aria-label="Carousel Tooltip"
              class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton sm-rounded inverse none undefined wvButtonRoundedIcon"
              data-accent-color=""
            >
              <div
                class="buttonContent iconLeft iconOnly"
              >
                <span
                  class="icon"
                  style="border-radius: 0;"
                >
                  <span
                    class="icon icon-info"
                  />
                </span>
              </div>
            </button>
            <div
              class="tooltipText"
            >
              <h5>
                1
              </h5>
              <p>
                Tooltip 1
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="next"
      >
        <button
          aria-label="Carousel Next Button"
          class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton xl tertiary none"
          data-accent-color=""
        >
          <div
            class="buttonContent iconLeft iconOnly"
          >
            <span
              class="icon"
              style="border-radius: 0;"
            >
              <span
                class="icon icon-next"
              />
            </span>
          </div>
        </button>
      </div>
      <div
        class="page"
      >
        1
        /
        5
      </div>
    </div>
  </div>
</body>
`;
